### PR TITLE
Pospow and SG triplanar fix

### PR DESCRIPTION
**Pull request by slunity**
### **Please read**
**PR workflow guidelines**
* SRP ABV will start automatically on Yamato when you open your PR
* Changes to docs and md files will **not** trigger ABV jobs 
* Consider making use of **draft PRs** if you are not 100% sure that your PR is ready for review
* ABV will restart if you add a new commit to a branch with an open PR (hence why you should consider using draft PRs)
* Adding [skip ci] (case insensitive) to the title of PRs will stop any jobs being trigger automatically - you will need to open Yamato and find your branch to run ABV
* You can also add [skip ci] to commit messages to prevent CI from running on that push
* Add [cancel old ci] to your commit message if you've made changes you want to test and no longer need the previous jobs

### Checklist for PR maker
- [x] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [x] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR.
- [ ] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Why is this PR needed, what hard problem is it solving/fixing?
Fixes a NaN issue in triplanar SG node, and make core PositivePow generally safer.
---
### Testing status

**Manual Tests**: What did you do?
- [x] Opened test project + Run graphic tests locally
- [ ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo + Prefab overrides + Alignment in Preset
- [ ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in developer mode, you have a button at end of resources that check the paths)
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)

**Yamato**: (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline

Any test projects to go with this to help reviewers?

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.shadergraph/Editor/Data/Nodes/UV/TriplanarNode.cs [hunk 1]
```diff
@@ -91,7 +91,7 @@ public virtual void GenerateNodeCode(ShaderStringBuilder sb, GenerationMode gene
                 // Whiteout blend method
                 // https://medium.com/@bgolus/normal-mapping-for-a-triplanar-shader-10bf39dca05a
                 case TextureType.Normal:
-                    sb.AppendLine("$precision3 {0}_Blend = max(pow(abs({1}), {2}), 0);"
```
Comment (slunity) [inline com.unity.shadergraph/Editor/Data/Nodes/UV/TriplanarNode.cs]: @sebastienlagarde Note how this is a false NaN guard, the compiler will ditch this because we compile with fast math and thus pow() is assumed to always return >= 0, ie no need for max( ? , 0)

Comment (alindmanUnity) [reply com.unity.shadergraph/Editor/Data/Nodes/UV/TriplanarNode.cs]: I see the function is in core, so I don't think it's an issue but to be safe did you check this against Universal as well? This node uses the same code for both pipelines so you need to test that it works safely in both.

File: com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl [hunk 2]
```diff
@@ -495,11 +495,58 @@ uint FastLog2(uint x)
...
 
-// Using pow often result to a warning like this
+// PositivePow: Same as pow(x,y) but considers x always positive and never exactly 0 such that
+// PositivePow(0,y) will numerically converge to 1 as y -> 0, including PositivePow(0,0) returning 1.
+//
+// Using pow often results in a warning like this
 // "pow(f, e) will not work for negative f, use abs(f) or conditionally handle negative values if you expect them"
-// PositivePow remove this warning when you know the value is positive or 0 and avoid inf/NAN.
-// Note: https://msdn.microsoft.com/en-us/library/windows/desktop/bb509636(v=vs.85).aspx pow(0, >0) == 0
-TEMPLATE_2_REAL(PositivePow, base, power, return pow(abs(base), power))
```
Comment (LandonTownsendUnity) [inline com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: This function is used in around 30 different files, giving this PR a large halo affect. If this is not meant to fix bugs with any other functionality using PositivePow, are we sure we want to make such a large change to fix Triplanar, instead of some other way?

File: com.unity.shadergraph/Editor/Data/Nodes/UV/TriplanarNode.cs (hunk 1 above)
Comment (slunity) [reply com.unity.shadergraph/Editor/Data/Nodes/UV/TriplanarNode.cs]: Good point, forgot to add the list of touched files also in my analysis in my commit comment.
In URP, /ShaderLibrary/MetaInput.hlsl uses it for an albedo boost from LightmapSettings (quite like a gamma) the rest is for gamma curve also (LutBuilderHdr and SMAA)

File: com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl (hunk 2 above)
Comment (slunity) [reply com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: Indeed it is used a bit everywhere, mostly for gamma curves, I've reviewed all cases and ran all HDRP_Tests and HDRP_RuntimeTests, passing. The additional max( ) should be extremely cheap. All that said, definitely has halo, I understand the reluctance. It just made

Comment (LandonTownsendUnity) [reply com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: Upon further inspection, this function is used in the majority of color conversion functions in color.hlsl (SRGB to linear, linear to gamma, etc) which means the halo effect is HUGE. I would strongly suggest not making a change to this function.

Comment (LandonTownsendUnity) [reply com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: Tests don't cover all edge cases. There are a good deal of edge cases I can think of where this may cause different behavior, especially in regards to black colors (0, 0, 0) being converted between color modes. Remember that with positivepow, when you have a base of 0 and a power of 1, you want the result to be 0, and this may not be the case anymore.
I think doing this as a function called SafePositivePow and only using that function in Triplanar and any other place that specifically needs NAN values corrected would be a much safer alternative.

Comment (slunity) [reply com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: Agree it is safer, will amend.

Comment (slunity) [reply com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: Done

File: com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl [hunk 3]
```diff
@@ -501,6 +501,57 @@ uint FastLog2(uint x)
...
+// 1) by leaving "y" free to get to 0, we get a NaNs
+// 2) the behavior of SafePositivePow() has more continuity when both x and y get closer together to 0, since
+// when x is assured to be positive non-zero, pow(x,x) -> 1 as x -> 0.
+//
+// TL;DR: SafePositivePow(x,y) avoids NaN and is safe for positive (x,y) including (x,y) == (0,0),
+//        but SafePositivePow(0, y) will return close to 1 as y -> 0, instead of 0, so watch out
+//        for behavior depending on pow(0, y) giving always 0, especially for 0 < y < 1.
+//
+// Ref: https://msdn.microsoft.com/en-us/library/windows/desktop/bb509636(v=vs.85).aspx
+float SafePositivePow(float base, float power)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: cant you use a macro function for this case?

TEMPLATE_2_FLOAT(SafePositivePow, base, power, return pow(max(abs(base), float(FLT_EPS)), power))

it is mean for this :)

Comment (sebastienlagarde) [reply com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: also why not a REAL version ?

TEMPLATE_2_REAL

Comment (slunity) [reply com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: I've fixed it, I had this sitting for a while but I think originally I mechanically did that for the constant vec type, which is useless obviously since scalars smear to larger types. (float I think it was to be consistent with FLT_EPS, and in the worst case if the platform/driver forces fp16 eg mobile, it reverted to the old function since FLT_EPS would become 0. The function is separate now anyway)

**Comment by slunity** [comment]
> Edit: I'm entering large values into the "blend" variable, I assume that this is causing divide by 0 
> errors instead of pow(0, 0) errors. Is this under the category of things we wanted to fix for this >node?
Ah yes, this is expected. 
I think just the name blend is not every informative and our documentation is not better (at all in fact) at explaining it to the user either. I'm thinking I should update the doc too. 
Blend is a "contrast enhancement" of the "normal selection" that is the basis of triplanar mapping.
In short, the more the blend exponent is high, the more the blending is biased towards the plane towards which the normal is point "more towards". 
In the limit, it should become a kind of "argmax", except like you noticed, it can quickly underflow the float precision and the normalization will make a divide to zero.

We can patch it and/or just warn the user. But I definitely think we need to update the doc too!

File: com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl [hunk 4]
```diff
@@ -501,6 +501,51 @@ uint FastLog2(uint x)
...
+// when x is assured to be positive non-zero, pow(x,x) -> 1 as x -> 0.
+//
+// TL;DR: SafePositivePow(x,y) avoids NaN and is safe for positive (x,y) including (x,y) == (0,0),
+//        but SafePositivePow(0, y) will return close to 1 as y -> 0, instead of 0, so watch out
+//        for behavior depending on pow(0, y) giving always 0, especially for 0 < y < 1.
+//
+// Ref: https://msdn.microsoft.com/en-us/library/windows/desktop/bb509636(v=vs.85).aspx
+TEMPLATE_2_REAL(SafePositivePow, base, power, return pow(max(abs(base), real(REAL_EPS)), power))
+
+// Helpers for making shadergraph functions consider precision spec through the same $precision token used for variable types
```
Comment (slunity) [inline com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl]: @sebastienlagarde Do you mind I added those here, so the triplanar node can seamlessly support the $precision without any new splicing keywords / system to get min/maxes etc.

Comment (sebastienlagarde) [reply com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl:537]: hey, yes you can do it, however you must chose, 
either:
TEMPLATE_2_REAL(SafePositivePow, base, power, return pow(max(abs(base), real(REAL_EPS)), power))

or both:
TEMPLATE_2_FLT(SafePositivePow_float, base, power, return pow(max(abs(base), float(FLT_EPS)), power))
TEMPLATE_2_HALF(SafePositivePow_half, base, power, return pow(max(abs(base), half(HALF_EPS)), power))

otherwise you will conflict. Haven't you get a compile error with function being redefined?

Comment (sebastienlagarde) [reply com.unity.render-pipelines.core/ShaderLibrary/Common.hlsl:537]: Ah sorry, misread :), missed the _float and _half.
ok then
